python/lsst/daf/butler/registries/postgresqlRegistry.py

```diff
@@ -0,0 +1,69 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("PostgreSqlRegistry", )
+
+import re
+import configparser
+import urllib.parse as urlparse
+import os
+
+from sqlalchemy import event
+from sqlalchemy import create_engine
+
+from lsst.daf.butler.core.config import Config
+from lsst.daf.butler.core.registry import RegistryConfig
+
+from .oracleRegistry import OracleRegistry
+from .sqlRegistry import SqlRegistryConfig
+
+class PostgreSqlRegistry(OracleRegistry):
+    """Registry backed by an PostgreSQL Amazon RDS service.
+
+    Parameters
+    ----------
+    config : `SqlRegistryConfig` or `str`
+        Load configuration
+    """
+
+    def __init__(self, registryConfig, schemaConfig, dimensionConfig, create=False,
+                 butlerRoot=None):
+        registryConfig = SqlRegistryConfig(registryConfig)
+        self.schemaConfig = schemaConfig
+        super().__init__(registryConfig, schemaConfig, dimensionConfig, create,
+                         butlerRoot=butlerRoot)
+
+    def _createEngine(self):
+        constr = self.config['db']
+
+        # if both username and pass are in the db string - all is well
+        # if username and pass are a nick that exists in a local conf file
+        # read username and password from there
+        parsed = urlparse.urlparse(constr)
+        if parsed.username and (parsed.password is None):
+            localconf = configparser.ConfigParser()
+            localconf.read(os.path.expanduser('~/.rds/credentials'))
```

Comment:
Except for the two review comments all others relate to `test_butler` code. As I mentioned in the PR - that code exists only as a quick way for anyone else that perhaps wants to take the `PostgreSqlRegistry` for a spin and because I thought it was a shame to completely "forget" about the steps I needed to take in test set-up. Considering the comment:

> ... happy to merge the first non-test code once we understand how credentials in external text files should be handled ...

I wonder wouldn't it be better to just totally remove the piece of test-code in a new commit where I sort out the authorization issues? This would leave the code in history (at least on my end) and simultaneously get rid of most of the complaints in this Draft-PR.

Reply:
You probably want an environment variable that you can set which points to the config file more than you want to put it in a well defined place. This sucks from a dev/prod database perspective if wanting to run code against both instances from the same home filesystem because you're going to be doing a bunch of symlink gymnastics. You can still have a default, which should probably be under `~/.lsst`, as it's not a Postgres config file (e.g. `.pgpass` file)

A postgres-specific solution would be to use postgres norms via env vars or a .pgpass file:
https://www.postgresql.org/docs/9.3/libpq-envars.html

Reply:
I read up on all of the suggestions made here and on the Slack channel yesterday. To summarize, the following suggestions were made:

* Postgres env vars 
  * not recommended by postgres manual as they say on some systems anyone can see any env variable.
* `.paf`, `.conf` file in `~/.rds`, `~/.postgres`, `~/.lsst` or `~/.butler` directory
  * the way its done now
  * I have a feeling (from @ktlim and @timj) that its `.paf` file using something built in the stack like the [`DBAuth` code](https://github.com/lsst/daf_persistence/blob/1e7c341a1a884cf06fc74a946a126c1f72f56360/src/DbAuth.cc) would be most preferred
  * but I would still like some clarification on where exactly should I put this: `~/.lsst` or `~/.butler`? `~/.rds`or `~/.postgres` seem to be not as popular
* `.pgpass` file
  * I can set the `PGPASSFILE` env var to point to `~/.lsst` or `~/.butler` directories so it works well with the file (above) approach
  * the format of entries in the file is given as `hostname:port:database:username:password` so it makes no sense to keep the current connection string format intact. I could change the current connection string to `postgresql://user:pass@hostname/dbname` and read the port and the password for the matching username, hostname and dbname from `.pgpass` file. This is also good as it simplifies the connection string.
  * but at some point if I want to connect to the DB using SQLAlchemy I still have to go there, read the file as plaintext and make substitutions in the connection string. So on face value it doesn't seem any different than the other file based approaches. 
  * The difference is only visible if from terminal one invokes a clean `:~$ psql` command as that is when postgres will natively go there and read the file if no user and pass env vars are set. If there's only one entry in the file it will connect to that directly otherwise `--username` `--host` and `--dbname` have to be given.
  * I'm not counting the env var as a win because that's easy to implement and depends on exactly the same questions as `.pgpass` does: when and where is the variable exported and to which location should it point to?
* vault + confd 
  * This requires a running `vault` server, which is simple enough I guess. I'm just confused about the general approach I guess: should there be one central vault, is everybody supposed to run their own, where is the vault in the AWS then, how do tests work... A lot of things I know not a lot about. 
  * Then the templatefiles need to be created by `confd`, which I guess could be a part of a startup script somewhere (where and when?)
  * but then again, every time I want to connect using SQLAlchemy I would have to run something like:
    ```
    ROOT_TOKEN=$(vault read -field id auth/token/lookup-self)

    confd -onetime -backend vault -node http://127.0.0.1:8200 \
          -auth-type token -auth-token $ROOT_TOKEN
    ```
    and would again have to read the files and make string substitutions to SQLAlchemy as was the case in all the options so far. Nobody can read the values in the files, even if they could access them, unless they luck out and look at them post token-substitutions? I guess it does get me more flexibility, especially if I run in deamon mode that will update the files if the back-end password changes. 

The dot files are plain text, but they are not public in the sense that they need to be in `600` or `400` permission set. All of the answers seem to boil down to reading the dotfile and making plain string substitutions. It was my understanding that the main problem was reading/keeping usernames and passwords floating about in plain text around the code. So I guess I'm confused where the security comes in to play in any of the suggestions, short of relying on OS privileges? (or maybe this is "the security", I don't know a lot about this stuff so someone needs to point it out to me)

The passwords are read prior to creating the connection string and passed on to SQLAlchemy as soon as possible (which then hides them again). This was the reasoning behind duplicating all the read-password code. If I write a function that will read and return username and password isn't that more dangerous than just doing it in place like it is? This is what, I though, I needed to avoid. 

I think I prefer the `.pgpass` file with `PGPASSFILE`, by default, pointing to one of the  `~/.lsst` or `~/.butler` directories by which is read by the LSST's `DBAuth` code approach the best. It gets me everything regular dotfile approach does but also integrates well with native `psql` commands. While we're at it, we can change the location of the AWS credentials file to be in the same default location as the rest of them by setting `AWS_SHARED_CREDENTIALS_FILE`. Any thoughts, comments and answers very welcomed! (especially, I think, input from @brianv0 might be very illuminating to me)

Thank you all for your help!

Reply:
I'm fine with a postgres file in a ~/.lsst directory. It's really important that you don't duplicate all the code for handling the password reading in multiple places. The more places you have it the more places where a mistake can be made. There is nothing wrong I can see with using function calls to obtain information and then inserting the answers into your connection string.  You should check that the file has the correct restricted permissions.

Reply:
Most of the work is focused on this comment. I have implemented a `ConnectionStringBuilder` class in `lsst.daf.butler.core.registry`. It constructs the connection string as generally as possible from the given config and `db-auth.paf` file. All tests pass (including the `RDSButlerTestCase` which was updated in previous commit but removed from the latest one). 

I think I might have over-extended the level of changes you requested but I wanted to avoid repeat of the experience with URI handling from before. I have followed the [SQLAlchemy manual](https://docs.sqlalchemy.org/en/13/core/engines.html) that cites [RFC-1738](https://www.ietf.org/rfc/rfc1738.txt) and all the default values are set equal to those [SQLAlchemy Dialect part](https://docs.sqlalchemy.org/en/13/dialects/index.html?highlight=dialects) of manual states. This should hopefully enable the construction of maximally general connection strings so all cases should be covered.

I was not 100% sure how Oracle connection string has been supplied but I suspect that they use conection strings without user and pass and then leave Oracle to work out the security. So I left that as an keyword option in `ConnectionStringBuilder.fromConfig`. 

I have left questions in form of comments in the `ConnectionStringBuilder` class that I can remove once I hear if you like this or not.